### ISSUE TITLE
Event management is a bottleneck with big datasets

### ISSUE THREAD

**Issue by Chosko** [Leaflet/Leaflet#5044]
I'm creating a lot of circle markers in a layer using a geoJson of 13000+ features.

```javascript
L.geoJson(data, {
    pointToLayer: function (feature, latlng) {
        return L.circleMarker(latlng);
    }
}).addTo(map);
```

this takes about 14 seconds in my browser (it's OK, it's a lot of data!). 
I profiled the CPU to understand what was taking so long, and I found that just this `for` takes 12 seconds.

https://github.com/Leaflet/Leaflet/blob/a97b87d0333fdc4b0c18d7b2ee9c15b393ae76ff/src/core/Events.js#L110

I think that using a different strategy to store the listeners, we could improve a lot the amount of data that Leaflet is able to insert in an acceptable amount of time.

**Comment by IvanSanchez** [comment]
Hi, great to hear that you find Leaflet useful!

However, this issue tracker is used for reporting bugs and discussing new features. For questions on using Leaflet, please use [gis.stackexchange.com](https://gis.stackexchange.com/questions/tagged/leaflet) or [stackoverflow](https://stackoverflow.com/questions/tagged/leaflet).

If you are *really* sure that this is a bug in leaflet, or a feature request, please [change the wording to make it look like a bug report](http://www.chiark.greenend.org.uk/~sgtatham/bugs.html).

**Comment by Chosko** [comment]
sorry, I hit CTRL+Enter and this posted the issue before it was ready.
It is a problem with Leaflet, just let me update the issue

**Comment by Chosko** [comment]
@IvanSanchez can you consider re-opening the issue please?

I'm working on this now, and should be able to send a PR soon

**Comment by IvanSanchez** [comment]
Attaching event listeners in such a way that they can be properly removed later has been an issue since the 1.0-beta1 refactors. If you are considering a PR, make sure you're running the unit tests.

The right way to attach event listeners to every member of a `LayerGroup`, and then use the `target` property of the event to see which layer originated the event. That way the event handlers are attached just once to the DOM elements, saving a lot of time.

**Comment by Chosko** [comment]
@IvanSanchez I again suggest to open this issue.
It's a serious performance bottleneck, and others applications may suffer for the same reason.
Leaving this closed prevents other people to help if they want to.

**Comment by IvanSanchez** [comment]
Can you publish a fiddle/codepen/[playground](https://playground-leaflet.rhcloud.com/) that displays this behaviour? Also, specify the Leaflet version and which browsers suffer from the most delay.

**Comment by Chosko** [comment]
Unfortunately I cannot provide a fiddle because I would publish private data, and I don't have enough time to mock so many features. These are all the informations I can write.

**Software versions**
Leaflet verion: 1.0.1
Browser: Chrome 54.0.2840.71 (64bit)
OS: macOS Sierra 12.0.49

**Geojson info**
size: 1.8MB
n. of features: 13417

**Code used to insert features as markers**

```javascript
L.geoJson(data, {
    pointToLayer: function (feature, latlng) {
        return L.circleMarker(latlng).bindTooltip(feature.properties.title);
    }
}).addTo(map);
```

where `data` is a variable containing the geoJson object

**CPU Profiling info**

[Here you can download a CPU profile](https://dl.dropboxusercontent.com/u/8771146/webdata/CPU-20161024T161132_last.cpuprofile) taken with Chrome. If you import it in Chrome you will see the following:

the `.addTo(map)` function call takes **17.51s** total time. (Image 1)
<img width="1280" alt="profile1" src="https://cloud.githubusercontent.com/assets/3297850/19650456/b49c04ce-9a09-11e6-9090-38ada8c561e2.png">

zooming the Chrome profiler timeline at random points on this function call, I can see that the deepest activation record is almost always the `._on` function (Image 2)
<img width="1272" alt="profile2" src="https://cloud.githubusercontent.com/assets/3297850/19650457/b49c7260-9a09-11e6-9baa-0da7046c04f8.png">

inside the `._on` function, the [for loop](https://github.com/Leaflet/Leaflet/blob/a97b87d0333fdc4b0c18d7b2ee9c15b393ae76ff/src/core/Events.js#L110)  itself  takes **13.397s** total time.
<img width="1275" alt="profile3" src="https://cloud.githubusercontent.com/assets/3297850/19650458/b49d40aa-9a09-11e6-8add-1a3db3e8d522.png">

This image shows the amount of features contained in the geojson
<img width="1277" alt="profile4" src="https://cloud.githubusercontent.com/assets/3297850/19650459/b49d8ad8-9a09-11e6-858e-4051a4704405.png">

When (and if) I find the time I will mock a geojson with that amount of data and create a fiddle.

**Comment by IvanSanchez** [comment]
> When (and if) I find the time I will mock a geojson with that amount of data and create a fiddle.

Done that for you: https://playground-leaflet.rhcloud.com/foq/1/edit?html,output (https://playground-leaflet.rhcloud.com/foq for output only), with 15k points.

That's taking about 3 secs in a chromium 53 in linux, and about 5 in a ffx  51.0a2 in linux.

The issue is *not* the tooltips, as https://playground-leaflet.rhcloud.com/vine/1/edit?html,output takes roughly the same time.

A bit of introspection shows that the only events that are attached in a loop are `L.Path`'s `zoomend` and `viewreset`. This happens with both the `L.Canvas` and `L.SVG` renderers. So theoretically those events could be handled by the `L.Renderer` and not each individual `L.Path`. 

While an optimization might be possible, I'm not sure if it would be a sane thing to do, given the complexity and error-proneness of past event refactors.

@perliedman, do you think this is worth reopening, as a performance improvement feature?

**Comment by perliedman** [comment]
It's tricky, like @IvanSanchez says. the closed PR #4709 more or less specifically attempts address this `for` loop (https://github.com/Leaflet/Leaflet/pull/4709/files#diff-0e79846ee2c407aa1ac0ccbfd8b15a44R125), but it was deemed that it performed worse in some important conditions, so we went with the current implementation.

I'm by no means satisfied with the current loop, it will perform bad in conditions like the one you describe, but it's tricky to find something that is also very fast for more standard cases _and_ performs correctly.

My own time and energy for diving into this again is limited, but if you have suggestions, they are welcome. Just be prepared for them to be turned down, since it's very easy to introduce subtle problems when fiddling with the events system, and we don't want to do that again.

**Comment by Chosko** [comment]
@IvanSanchez, @perliedman thanks for your support! 
This issue is critical in my application, so I'll try to find a workaround/solution for my specific case. And if I find a solution which is globally good in terms of performance and well tested I'll send you a PR. So, thanks for now.

However, I still don't get all this reluctance in opening the issue. I see on this repository lots issues, a few of them without enough data to understand the problem or others with 'needs more info' but still open. This one has been closed 3 minutes after I wrote it, before having the possibility to be explained and without the possibility to be discussed.
I understand that you may have your interest in keeping low the number of issues, but the problem explained here *is real* and *well defined* and I'm sure there may be people who have the same problem all over the Internet and they may want to help (like I'm trying to do). 
Adding 'needs discussion' while keeping this closed makes no sense to me. Sorry for my insistence, but I'm the first who is interested in seeing if others will come with a possible solution.

However, you have the last word (that's right) and I won't bother again.

**Comment by perliedman** [comment]
Hm, I actually thought I reopened it yesterday, but apparently must have missed the button. Sorry about that.

**Comment by perliedman** [comment]
After a second look at #4709, it appears it tries to accomplish more, which might be the reason we finally had to to abandon the idea.

One thing I thought about would be to implement some conditional logic, like when the number of listeners exceeds a certain number (say 20), we switch to also keep track of listeners in a hash, so that we don't have to loop through them, but just look them up in the hash. This would keep the performance as it is now for normal cases, where looping a few listeners is quick, but also give good performance for extremes where we have thousands of listeners.

It would also be really good if we found a way to better benchmark the performance of the event system for a couple of realistic scenarios.

**Comment by Chosko** [comment]
Yes, it's a good idea. I'll try with that. Thanks.

**Comment by Chosko** [comment]
I made the performance tests for the pull request #5050 which should fix the problem

Before the PR
https://playground-leaflet.rhcloud.com/foq/1/edit?html,output
Time elapsed: 4371.655000000002msec

After the PR
https://playground-leaflet.rhcloud.com/bal/edit?html,output
Time elapsed: 1089.2650000000003msec

**Comment by Chosko** [comment]
However, I'd like to see some tests on normal cases with a few listeners, just to see if nothing bad happens....

**Comment by IvanSanchez** [comment]
Playground with #5054: https://playground-leaflet.rhcloud.com/giw/edit?html,output
Time: ~300msec

**Comment by IvanSanchez** [comment]
Closed by #5054
